### PR TITLE
SONARIAC-1435 Generate Parser and Visitor with ANTLR for properties file

### PR DESCRIPTION
**Pull request by mstachniuk**

### CODE REVIEW

File: iac-extensions/spring-config/build.gradle.kts [hunk 1]
```diff
@@ -1,11 +1,14 @@
 plugins {
+    antlr
     id("org.sonarsource.iac.code-style-convention")
     id("org.sonarsource.iac.java-conventions")
 }
 
 description = "SonarSource IaC Analyzer :: Extensions :: Spring Config"
 
 dependencies {
+    antlr("org.antlr:antlr4:4.13.1")
```
Comment (petertrr) [inline iac-extensions/spring-config/build.gradle.kts:10]: Please add this to `libs.versions.toml`

File: iac-extensions/spring-config/build.gradle.kts [hunk 2]
```diff
@@ -15,3 +18,17 @@ dependencies {
     testImplementation(libs.sonar.analyzer.test.commons)
     testImplementation(testFixtures(project(":iac-common")))
 }
+
+tasks.generateGrammarSource {
+    maxHeapSize = "64m"
+    arguments = arguments + listOf("-visitor", "-long-messages", "-no-listener", "-package", "org.sonar.iac.springconfig.parser.properties")
+    // Due to bug in ANTLR Gradle plugin https://stackoverflow.com/a/49388412
```
Comment (petertrr) [inline iac-extensions/spring-config/build.gradle.kts:25]: I don't get what the bug is. Is it that ANTLR source is not in a package but we want the generated file to be? Then I wouldn't call it a bug and rephrase the comment like `antrl plugin requires grammar file to be placed without package in path, but we want the generated file to follow standard Java convention`

File: iac-extensions/spring-config/build.gradle.kts [hunk 3]
```diff
@@ -15,3 +18,17 @@ dependencies {
...
+    outputDirectory =
+        project.layout.buildDirectory.dir(
+            "generated-src/antlr/main/org/sonar/iac/springconfig/parser/properties/"
```
Comment (petertrr) [inline iac-extensions/spring-config/build.gradle.kts:28]: Can it be simple `generated-src/main` and not `generated-src/antlr/main`?

File: build-logic/src/main/kotlin/org.sonarsource.iac.code-style-convention.gradle.kts [hunk 4]
```diff
@@ -14,7 +14,7 @@ spotless {
                 )
             )
         licenseHeaderFile(rootProject.file("LICENSE_HEADER")).updateYearWithLatest(true)
-        targetExclude("*/generated-sources/**")
+        targetExclude("*/generated-sources/**", "*/generated-src/**")
```
Comment (petertrr) [inline build-logic/src/main/kotlin/org.sonarsource.iac.code-style-convention.gradle.kts:17]: Why not configure it to output to `generated-sources` as well?

File: iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+      "visitRow mail.defaultRecipients[1]=owner@mail.com<EOF>",
+      "visitLine mail.defaultRecipients[1]=owner@mail.com<EOF>",
+      "visitKey mail.defaultRecipients[1]",
+      "visitKey owner@mail.com",
+      "visitEol <EOF>");
+  }
+
+  @Test
+  void shouldParseDifferentNewLines() {
+    var code = "foo1=bar1\nfoo2=bar2\rfoo3=bar3\r\nfoo4=4\n";
```
Comment (petertrr) [inline iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java:131]: I would also suggest adding tests with empty lines between properties to see how many trailing newlines are included in the text. And generally, I would suggest adding more examples based on the repo with the grammar (multiline values, etc.)

**Review by petertrr — COMMENTED** [review]
Good progress, please see my questions

File: iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java (hunk 5 above)
Comment (jonas-wielage-sonarsource) [reply iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java:131]: Specifically this file contains a lot of cases: 
https://github.com/antlr/grammars-v4/blob/master/properties/examples/example1.txt

File: build-logic/src/main/kotlin/org.sonarsource.iac.code-style-convention.gradle.kts (hunk 4 above)
Comment (mstachniuk) [reply build-logic/src/main/kotlin/org.sonarsource.iac.code-style-convention.gradle.kts:17]: The `generated-src` was a default. We have to anyway configure it, so I'm ok to change it.

File: iac-extensions/spring-config/build.gradle.kts (hunk 3 above)
Comment (mstachniuk) [reply iac-extensions/spring-config/build.gradle.kts:28]: It has to start with `generated-src/antlr/main` as this prefix is used as output to the task. Not following this convention requires additional configuration of this task or `sourcesJar`.

File: iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java [hunk 6]
```diff
@@ -0,0 +1,792 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import org.antlr.v4.runtime.tree.ParseTree;
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class PropertiesParserBaseVisitorTest {
```
Comment (petertrr) [inline iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java:31]: Could you maybe also add one test for text ranges to demonstrate that what is provided is reasonable? The visitor can collect this data in another list, and in a test also string matching can be done.

**Review by petertrr — APPROVED** [review]
I don't have any major questions, so it LGTM

File: iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java [hunk 7]
```diff
@@ -0,0 +1,792 @@
...
+      "visitKey bar3",
+      "visitEol \\u2029");
+  }
+
+  @Test
+  void shouldAcceptAllCharactersInComments() {
+    var code = """
+      #comment 1 ! # : =
+      !comment 2 # aa
+      foo2=bar2""";
```
[the remark this answers is not in the document]
Comment (mstachniuk) [reply iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java:346]: Sure I will add it. Trailing comments are not allowed - they will be a part of the value.
